<!--
AMIR 740をROS Controlで制御するサンプルのlaunch fileです。
実機を動かす場合は、先にROSとの接続をしてください。（詳しくは、ROS制御モード取扱説明書を確認してください）

RViz上(MoveItのインターフェイス)でAMIR 740の目標ポーズを指定し、
モーションプラニングが行われて、その軌道を実行可能です。

起動コマンド： roslaunch amir_control amir_control_HW.launch

-->

<launch>
 <!-- RViz関連 -->
    <!--パラメータサーバにURDFをロード -->
    <param name="robot_description" command="$(find xacro)/xacro '$(find amir_description)/urdf/amir.xacro'" />

    <!-- Rvizに表示  -->
    <node name="rviz" pkg="rviz" type="rviz" args="-d $(find amir_control)/basic.rviz"/>

<!-- コントローラ関連 -->
    <!-- コントローラの設定をロード  -->
    <rosparam file="$(find amir_control)/config/amir_controllers.yaml" command="load"/>

    <!-- hardware interfaceをロード -->
    <node name="amir_hw_main" pkg="amir_control" type="amir_hw_main" output="screen"/> 

    <!-- プランニングシーンをロード -->
    <node name="planning_scene_primatives" pkg="amir_motion" type="planning_scene_primatives" output="screen"/> 

    <!-- controller managerをロード  -->
    <node name="ros_control_controller_manager" pkg="controller_manager" type="controller_manager" respawn="false"
        output="screen"
        args="spawn joint_state_controller arm_velocity_trajectory_controller gripper_velocity_trajectory_controller" />

    <!-- joint_statesを/tfに変換 -->
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"/>

    <!-- 疑似的なシミュレータ（モーション確認用）
    （実機を動かす時、これをコメントアウトしてください） -->
    <!-- <node name="amir_sim_echo" pkg="amir_control" type="amir_sim_echo"/> -->

 <!-- MoveIt関連 -->
    <!-- パラメータサーバに必要なファイル（URDF, SRDF, .yaml configurationsなど）をロード -->
    <include file="$(find amir_moveit_config)/launch/planning_context.launch">
    <arg name="load_robot_description" value="true"/>
    </include>

    <!-- follow_joint_trajectoryをremapする -->
    <remap from="/arm_controller/follow_joint_trajectory" to="/arm_velocity_trajectory_controller/follow_joint_trajectory"/>
    <remap from="/gripper_controller/follow_joint_trajectory" to="/gripper_velocity_trajectory_controller/follow_joint_trajectory"/>

    <!-- MoveItを実行のため -->
    <include file="$(find amir_moveit_config)/launch/move_group.launch">
    <arg name="allow_trajectory_execution" value="true"/>
    <arg name="info" value="true"/>
    </include>
    
    <!-- 初期位置のズレのエラー防止用 -->
    <param name="move_group/trajectory_execution/allowed_start_tolerance" value="0.0" />
    <param name="move_group/trajectory_execution/allowed_execution_duration_scaling" value="3.0" />

</launch>